File: src/main/java/ch/uzh/ifi/access/projections/CourseOverview.java
package ch.uzh.ifi.access.projections;

import ch.uzh.ifi.access.model.Course;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.data.rest.core.config.Projection;

import java.time.LocalDate;
import java.util.List;

@Projection(types = {Course.class})
public interface CourseOverview {

    Long getId();

    String getUrl();

    String getTitle();

    String getUniversity();

    String getSemester();

    String getAvatar();

    LocalDate getStartDate();

    LocalDate getEndDate();

    String getDuration();

    String getDescription();

    @Value("#{@courseService.calculateCoursePoints(target.assignments, null)}")
    Double getPoints();

    @Value("#{@courseService.getMaxPoints(target.url)}")
    Double getMaxPoints();

    @Value("#{@activityRegistry.getOnlineCount(target.url)}")
    Long getOnlineCount();

    Long getStudentsCount();

    @Value("#{@courseService.getTeamMembers(target.supervisors)}")
    List<MemberOverview> getSupervisors();

    @Value("#{@courseService.getTeamMembers(target.assistants)}")
    List<MemberOverview> getAssistants();
}

File: src/main/kotlin/ch/uzh/ifi/access/projections/CourseOverview.kt
package ch.uzh.ifi.access.projections

import ch.uzh.ifi.access.model.Course
import ch.uzh.ifi.access.model.CourseInformation
import ch.uzh.ifi.access.model.constants.Visibility
import org.springframework.beans.factory.annotation.Value
import org.springframework.data.rest.core.config.Projection
import java.time.LocalDateTime

@Projection(types = [Course::class])
interface CourseOverview {
    val id: Long?
    val slug: String?
    val logo: String?
    val information: Map<String?, CourseInformation?>?
    val defaultVisibility: Visibility
    val overrideVisibility: Visibility?
    val overrideStart: LocalDateTime?
    val overrideEnd: LocalDateTime?

    @get:Value("#{@courseService.calculateCoursePoints(target.slug)}")
    val points: Double?

    @get:Value("#{@pointsService.getMaxPoints(target.slug)}")
    val maxPoints: Double?

    @get:Value("#{@visitQueueService.getRecentlyActiveCount(target.slug)}")
    val onlineCount: Long?
    val participantCount: Long?

    @get:Value("#{@courseService.getTeamMembers(target.supervisors)}")
    val supervisors: Set<MemberOverview>?

    @get:Value("#{@courseService.getTeamMembers(target.assistants)}")
    val assistants: Set<MemberOverview>?
}
